feat(episodes): add episode availability tracking and sync

0xSysR3ll opened the pull request:
<!--
    Please read contributing guide before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

## Description

<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

This allows seerr to track the availability status of individual episodes, enabling better status reporting for partially available seasons.

- Fixes #270 

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

- Complete season except one episode available.
- Correctly show the "available" badge for each episode except for the missing one.

## Screenshots / Logs (if applicable)

### Settings
<img width="472" height="85" alt="image" src="https://github.com/user-attachments/assets/2042feaa-ae3c-4f9b-99ee-8ff6cf0699dc" />

### Result
<img width="1281" height="592" alt="image" src="https://github.com/user-attachments/assets/caf8bc3d-fb1d-4fe8-b471-9e1b20c04fec" />

### If TVDB not set
<img width="1201" height="174" alt="image" src="https://github.com/user-attachments/assets/e47e3816-dde4-45ff-8b48-a837f281d283" />

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read and followed the contribution [guidelines](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md).
- [x] Disclosed any use of AI (see our [policy](https://github.com/seerr-team/seerr/blob/develop/CONTRIBUTING.md#ai-assistance-notice))
- [x] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] Successful build `pnpm build`
- [x] Translation keys `pnpm i18n:extract`
- [x] Database migration (if required)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Toggle to enable episode availability tracking in Settings with helper text and localization
  * Episode availability badges shown on TV season/episode views when enabled
  * Public settings now expose episode availability and metadata provider options

* **Chores**
  * Database and models updated to store and surface per-episode availability
  * Background sync updated to fetch, cache and persist per-episode availability
  * API responses include availability mapping for seasons

* **Migrations**
  * Added migration to create episode table for availability storage
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

0xSysR3ll committed:
feat(episodes): add episode availability tracking and sync

This allows Jellyseerr to track the availability status of individual
episodes, enabling better status reporting for partially available seasons.

fallenbagel commented:
This will not work properly when theres a mismatch between tvdb and tmdb right? How are you handling that

0xSysR3ll commented:
> This will not work properly when theres a mismatch between tvdb and tmdb right? Thats why our original plan was to wait till tvdb pr is merged until this feature. 
> 
> Unless you have some other ideas? (Haven't gone through the code yet)

You’re absolutely right about the TVDB/TMDB mismatch concern - I considered that during implementation. Instead of relying on direct TMDB ID mapping, I leveraged the existing Sonarr connection via externalServiceId and match episodes based on season and episode numbers. This approach works well for the majority of shows where episode numbering aligns across databases.

That said, I recognize there will be edge cases where TVDB and TMDB have inconsistent episode structures. I saw this as a pragmatic first step that covers most use cases.

I thought about a few options moving forward:
	•	Merge this as-is to support the majority of shows now.
	•	Hold off until the TVDB PR lands to enable more precise mapping.
	•	Make this behavior optional until we have more robust TVDB integration.

Wdyt ?

fallenbagel commented:
> Hold off until the TVDB PR lands to enable more precise mapping.

I think this would be the better idea. Or it will cause an influx of support requests saying why is this episode not available when it is etc etc

0xSysR3ll committed:
chore(tv): removed the wrong parameter.

The PR should not fix the issue reported in recent tvdb PR.

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
chore(tv): removed the wrong parameter.

The PR should not fix the issue reported in recent tvdb PR.

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll commented:
> The merge of the tvdb is done

Now it is yes

gauthier-th commented:
Following up on the previous comments about episode mismatch with TMDB:
Do you think it would be a good idea to be able to enable this feature ONLY when TVDB is configured as metadata provider?
As said above, people will not understand why there are a lot of mismatch when they're using TMDB (the default metadata provider) and will constantly ask why there is a "bug".

0xSysR3ll committed:
feat(settings): make the feature optionnal

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

gauthier-th commented on src/components/TvDetails/Season/index.tsx:
```diff
@@ -52,6 +54,13 @@ const Season = ({ seasonNumber, tvId }: SeasonProps) => {
                     {episode.airDate && (
                       <AirDateBadge airDate={episode.airDate} />
                     )}
+                    <Badge badgeType={episode.available ? 'success' : 'danger'}>
+                      {intl.formatMessage(
+                        episode.available
+                          ? globalMessages.available
+                          : globalMessages.unavailable
+                      )}
+                    </Badge>
```
This badge will show even if the metadata provider is not TVDB right?
Shouldn't we display it only for TVDB? It doesn't really make sense to display it otherwise.

0xSysR3ll replied:
Yep. You're right, I totally forgot about the frontend!

0xSysR3ll commented on cypress/config/settings.cypress.json:
```diff
@@ -201,4 +202,4 @@
     }
   }
-}
+}
```
Shouldn't prettier handle this ?

0xSysR3ll committed:
fix(ui): only display availability if provider is tvdb

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
fix(api): only mark episode as available if provider is tvdb

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
fix(cypress): add missing newline

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
fix(tv): rely on provider type instead of setting

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll commented on src/components/TvDetails/Season/index.tsx:
```diff
@@ -30,6 +39,15 @@ const Season = ({ seasonNumber, tvId }: SeasonProps) => {
+  const isAnime = tvData?.keywords.some(
+    (keyword) => keyword.id === ANIME_KEYWORD_ID
+  );
+  const isTvdbProvider = metadataSettings
```
I mean to check if tvdb is set as the provider. Not sure my implementation is the best.

0xSysR3ll replied:
> WDYT of that:
> 
> * Include the `metadataSettings` in the public settings, so you can then use the `useSettings()` hooks without having to fetch the metadata settings every time you open the Disclosure containing the episodes.
> * Add a `isAnime` parameter to the `<Season />` component and use it in `src/components/TvDetails/index.tsx` to avoid one more API call to fetch the TV show detail.

TBH, I initially thought about doing this but assumed it might be out of scope for this PR. If you’re fine with it, I can go ahead and include it.

0xSysR3ll committed:
feat(settings): add metadata settings for TV and anime

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
refactor(tv): simplify episode availability checks

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
fix(settings): missing metadata settings

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
feat(availability): implement episode caching

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
refactor(season): make episodes property optional and remove eager loading

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
refactor(api): remove getEpisodesBySeriesId method and update availabilitySync to use getEpisodes

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
chore: apply linting

0xSysR3ll committed:
refactor(tv): change getSettings to synchronous call

0xSysR3ll committed:
refactor(availability): enhance episode tracking logic in availability sync

0xSysR3ll committed:
chore: restore old changes

0xSysR3ll committed:
chore: reapply linting

gauthier-th reviewed (changes requested):
From what I'm seeing TypeORM migrations are 6+ months old. Please re-generate them.

0xSysR3ll committed:
feat(migration): add episode table migrations

gauthier-th commented on server/entity/Season.ts:
```diff
@@ -29,6 +31,11 @@ class Season {
+  @OneToMany(() => Episode, (episode) => episode.season, {
+    cascade: true,
+  })
+  public episodes?: Episode[];
```
Same here, you should add an index.

0xSysR3ll replied:
Mmm, if we add an index on Episode's entity I'm not sure what we'd gain from an index on Season's side?

fallenbagel replied:
I would say an index is needed here since the episode index covers the episode.seasonId FK, which should help when joining episodes to a season. The Season index covers season.mediaId, which helps when joining seasons to a media. Two different things on two different columnts. One doesn't help the other and indexes are fairly low cost in writing so should help a lot with performance when reading.

Though make sure it is added on the ManyToOne episode entity side where the `seasonId` FK column actually lives, not on the OneToMany here. OneToMany doesn't own a column so index here wouldn't really do anything

gauthier-th replied:
> If we already add an index on Episode's entity I'm not sure what we'd gain from an index on Season's side?

Yeah my bad, not needed. We already have the index on `season.mediaId` from #2461, and you added the index on `episode.season`, so it should be ok. Just one thing: could you move the `@Index` of `Episode` just before the `season` field? I find this more clear than having it at the beginning of the file.

0xSysR3ll replied:
Nitpick but I understand :p

0xSysR3ll committed:
feat(episode): add index on season

0xSysR3ll committed:
feat(settings): add warning for episode availability without TVDB set

0xSysR3ll committed:
refactor(migration): replace old episode table migration with to include season index

0xSysR3ll committed:
refactor(episode): move index on ManyToOne relationship with Season

0xSysR3ll committed:
Merge branch 'develop' into feat-episode-availability

gauthier-th commented on server/entity/Episode.ts:
```diff
@@ -0,0 +1,50 @@
+  @DbAwareColumn({
+    type: 'datetime',
+    default: () => 'CURRENT_TIMESTAMP',
+    onUpdate: 'CURRENT_TIMESTAMP',
```
`onUpdate` should not be used. See #2823

0xSysR3ll committed:
feat(episodes): add episode availability tracking and sync

This allows Jellyseerr to track the availability status of individual
episodes, enabling better status reporting for partially available seasons.

0xSysR3ll committed:
feat(settings): make the feature optionnal

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
fix(ui): only display availability if provider is tvdb

0xSysR3ll committed:
fix(api): only mark episode as available if provider is tvdb

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
fix(cypress): add missing newline

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
fix(tv): rely on provider type instead of setting

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
feat(settings): add metadata settings for TV and anime

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
refactor(tv): simplify episode availability checks

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
fix(settings): missing metadata settings

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
feat(availability): implement episode caching

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
refactor(season): make episodes property optional and remove eager loading

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
refactor(api): remove getEpisodesBySeriesId method and update availabilitySync to use getEpisodes

Signed-off-by: 0xsysr3ll <0xsysr3ll@pm.me>

0xSysR3ll committed:
chore: apply linting

0xSysR3ll committed:
refactor(tv): change getSettings to synchronous call

0xSysR3ll committed:
refactor(availability): enhance episode tracking logic in availability sync

0xSysR3ll committed:
chore: restore old changes

0xSysR3ll committed:
chore: reapply linting

0xSysR3ll committed:
feat(migration): add episode table migrations

0xSysR3ll committed:
feat(episode): add index on season

0xSysR3ll committed:
feat(settings): add warning for episode availability without TVDB set

0xSysR3ll committed:
refactor(migration): replace old episode table migration with to include season index

0xSysR3ll committed:
refactor(episode): move index on ManyToOne relationship with Season

0xSysR3ll committed:
refactor(Episode): we shouldn't use onUpdate

0xSysR3ll committed:
feat: update migrations

0xSysR3ll committed:
chore: remove old migrations

gauthier-th reviewed (dismissed):
LGTM. A rebase with a new generation of the migration will be needed before we merge this though.

0xSysR3ll committed:
chore: upgrade migrations

matt-pinder commented:
Looking forward to this functionality 🙌

Thankyou

gauthier-th reviewed (changes requested):
Due to #2712 being merged with db migration, can you please re-generate yours? And then it should be good 😅

0xSysR3ll committed:
chore: update db migrations

gauthier-th reviewed (dismissed):
LGTM. Can you rebase on develop so I can create a preview tag?

fallenbagel commented on server/routes/tv.ts:
```diff
@@ -84,7 +85,39 @@ tvRoutes.get('/:id/season/:seasonNumber', async (req, res, next) => {
     });
 
-    return res.status(200).json(mapSeasonWithEpisodes(season));
+    const availableMap: Record<number, boolean> = {};
```
`availableMap` is declared as `{}` unconditionally and only populated inside `if (shouldTrackEpisodes)` below. Since `mapEpisodeResult` checks `availableMap ? ... : undefined`, an empty object is still truthy, so wouldn't this mean when episode tracking is disabled (or TVDB isn't configured), every episode gets `available: false` instead of `available: undefined`?

I think we should declare it as a `let availableMap: Record<number, boolean> | undefined;` and only assigning `{}` inside the `shouldTrackEpisodes` block. Wdyt? @0xSysR3ll @gauthier-th 

Also available isn't documented in seerr-api.yml's Episode schema. Please add as an optional boolean.

fallenbagel commented on server/routes/tv.ts:
```diff
@@ -84,7 +85,39 @@ tvRoutes.get('/:id/season/:seasonNumber', async (req, res, next) => {
+            if (dbSeason.episodes) {
+              for (const episode of dbSeason.episodes) {
+                availableMap[episode.episodeNumber] =
+                  episode.status === MediaStatus.AVAILABLE;
```
`episodeNumber` here comes from Sonarr's `getEpisodes` (season-relative), while `episode_number` in the AVAILABLE branch above comes from TMDB.

In practice these should align for TVDB-backed anime since both are season-relative, but there's no explicit guard if a mismatch ever occurs meaning episodes could silently show wrong availability rather than erroring. 

This could cause issues in the the future when debugging if anime availability badges look off.

fallenbagel commented on src/components/Settings/SettingsMain/index.tsx:
```diff
@@ -70,6 +72,11 @@ const messages = defineMessages('components.Settings.SettingsMain', {
   enableSpecialEpisodes: 'Allow Special Episodes Requests',
+  enableEpisodeAvailability: 'Enable Episode Availability Tracking',
+  enableEpisodeAvailabilityTip:
+    'Track individual episode availability status (requires TVDB as metadata provider for TV shows or anime)',
```
```suggestion
    'Track individual episode availability status (requires TVDB as metadata provider for either TV shows or anime)',
```

0xSysR3ll committed:
fix: omit episode availability when tracking is disabled

0xSysR3ll committed:
fix(tv): handle episode number mismatches in availability checks

0xSysR3ll committed:
fix: clarify tracking message in settings

fallenbagel commented on server/entity/Media.ts:
```diff
@@ -69,9 +69,27 @@ class Media {
+        relations.seasons = {
+          episodes: true,
+        };
+      }
```
This isn't gated on `enableEpisodeAvailability`, so every TV fetch through `getMedia()` now eager-loads every episode of every season. 

The only caller that needs episode data is the season route, which already fetches media separately and gates it behind `shouldTrackEpisodes`. The main `/:id` route calls `getMedia()` unconditionally too, and `mapTvDetails` never reads episode data. So this adds a heavier join to every show page load even when the feature is off. Can this be conditioned on the setting, or moved out of this shared method?

0xSysR3ll replied:
Makes sense, I'll move this from here.

fallenbagel commented on server/lib/availabilitySync.ts:
```diff
@@ -858,6 +886,83 @@ class AvailabilitySync {
+              const existingEpisodes = await episodeRepository.find({
+                where: { season: { id: season.id } },
+                relations: ['season'],
+              });
```
This does a DB read (`episodeRepository.find`) per season, per media item, on every sync cycle. Given this runs against potentially thousands of shows, might be worth batching this.

Like fetch all episodes for the media item once (or once per sync pass) instead of re-querying per season.

0xSysR3ll replied:
Though I had already done this. Good catch.

fallenbagel commented on server/lib/availabilitySync.ts:
This job only ever downgrades already **AVAILABLE**/**PARTIALLY_AVAILABLE** items when content's **gone** (`loadAvailableMediaPaginated`'s query, and every other status write in this file goes toward `DELETED`, never `AVAILABLE`). **The new episode code goes the other way**. It's the only place `Episode.status` gets set at all, and it sets `AVAILABLE` here. Scanners (`baseScanner.ts` + `jellyfin`/`plex`/`sonarr`/`radarr`) are what establish season/media availability everywhere else and aren't touched by this PR. 

Episode availability should be established there (the actual scanners), in sync with when the season itself is scanned, instead of depending on this job's sweep cadence. Also, using `UNKNOWN` instead of `DELETED` for "no file" is inconsistent with how this file treats everything else that's gone missing (soft removal).

In addition:

https://github.com/0xSysR3ll/jellyseerr/blob/5a3e96f2d5e616dc7656f003d0cb0f04dbee0f02/server/lib/availabilitySync.ts#L60-L63

This ORs tv/anime globally and with only `anime: TVDB` set, non-anime shows get Sonarr episode calls and Episode row writes here too. The mismatch check in https://github.com/seerr-team/seerr/pull/1671/commits/a91506da477ae216cb8923c041d5b53242914dd1 will suppress the badge for most of these anyway, so does the write side need the OR (ignoring the fact that episodes shouldn't be made available in file)?

0xSysR3ll replied:
Hum, would it be merge-blocking to not move episode availability into scanners ? Cause I can do it but this means a bigger PR. Unless I do it in a stacked PR. Wdyt ?

fallenbagel replied:
> Hum, would it be merge-blocking to not move episode availability into scanners ? Cause I can do it but this means a bigger PR. Unless I do it in a stacked PR. Wdyt ?

Stacked PR is okay. We can merge from the top stack instead of this once thats done so both lands together. But remove this logic entirely from availability sync. It should never promote any media. Separation of concerns

0xSysR3ll replied:
Sounds good to me then, will push a stacked PR.

0xSysR3ll committed:
refactor: load episode relations only in the season route

0xSysR3ll committed:
fix: keep episode availability out of availability sync
